### PR TITLE
Update library with linkable tree

### PR DESCRIPTION
**Pull request by drewstone**
Things left to do:
- [ ] Update the circuits with the set membership proof, add the roots and length parameters to the circuits.
- [ ] Add the meta verifier that delegates to differently sized (by length) verifiers.
- [ ] Update the existing tests to pass.

### CODE REVIEW

File: test/Semaphore.ts [hunk 1]
```diff
@@ -136,29 +136,41 @@ describe("Semaphore", () => {
         })
 
         it("Should not verify a proof if the group does not exist", async () => {
-            const transaction = contract.verifyProof(10, bytes32Signal, 0, 0, [0, 0, 0, 0, 0, 0, 0, 0])
+            const roots = await contract.getRoot(groupId);
```
Comment (drewstone) [inline test/Semaphore.ts:139]: Right now this is wrong, it should in fact be the concatenated root set, not just the root of the group's tree on this chain.

File: contracts/extensions/SemaphoreWhistleblowing.sol [hunk 2]
```diff
@@ -75,13 +75,14 @@ contract SemaphoreWhistleblowing is ISemaphoreWhistleblowing, SemaphoreCore, Sem
         bytes32 leak,
         uint256 nullifierHash,
         uint256 entityId,
+        bytes calldata roots,
```
Comment (drewstone) [inline contracts/extensions/SemaphoreWhistleblowing.sol:78]: For these functions (on the derivative contracts) we still need to verify that these roots are in the history.

**Review by drewstone — COMMENTED** [review]
@semaraugusto also fix up the style/tabs if you can!

**Comment by semaraugusto** [comment]
Should be ready for review:

- [x] Update the circuits with the set membership proof, add the roots and length parameters to the circuits.
- [x] Add the meta verifier that delegates to differently sized (by length) verifiers.
- [x] Update the existing tests to pass.

File: contracts/Semaphore.sol [hunk 3]
```diff
@@ -2,18 +2,22 @@
...
+// import "./interfaces/IVerifier.sol";
+import "./verifiers/SemaphoreVerifier.sol";
 import "./base/SemaphoreCore.sol";
 import "./base/SemaphoreGroups.sol";
 
 /// @title Semaphore
 contract Semaphore is ISemaphore, SemaphoreCore, SemaphoreGroups {
     /// @dev Gets a tree depth and returns its verifier address.
-    mapping(uint8 => IVerifier) public verifiers;
+    mapping(uint8 => SemaphoreVerifier) public verifiers;
```
Comment (drewstone) [inline contracts/Semaphore.sol:13]: I think we need a double map here? Tree depth + max edges?

Comment (drewstone) [reply contracts/Semaphore.sol:13]: That is, if we are to support all sorts of configurations

File: contracts/Semaphore.sol [hunk 4]
```diff
@@ -82,18 +87,22 @@ contract Semaphore is ISemaphore, SemaphoreCore, SemaphoreGroups {
...
 
         if (depth == 0) {
             revert Semaphore__GroupDoesNotExist();
         }
 
-        IVerifier verifier = verifiers[depth];
+        SemaphoreVerifier verifier = verifiers[depth];
 
-        _verifyProof(signal, root, nullifierHash, externalNullifier, proof, verifier);
+        // _verifyProof(signal, nullifierHash, externalNullifier, roots, proof, verifier, 1);
```
Comment (drewstone) [inline contracts/Semaphore.sol:104]: Remove unneeded comment.

File: contracts/base/SemaphoreCore.sol [hunk 5]
```diff
@@ -2,46 +2,66 @@
 pragma solidity ^0.8.4;
 
 import "../interfaces/ISemaphoreCore.sol";
-import "../interfaces/IVerifier.sol";
+// import "../interfaces/IVerifier.sol";
```
Comment (drewstone) [inline contracts/base/SemaphoreCore.sol:5]: Remove if unneeded.

File: contracts/base/SemaphoreCore.sol [hunk 6]
```diff
@@ -2,46 +2,66 @@
...
     /// @dev Asserts that no nullifier already exists and if the zero-knowledge proof is valid.
     /// Otherwise it reverts.
     /// @param signal: Semaphore signal.
-    /// @param root: Root of the Merkle tree.
+    /// @param roots: Roots of the Merkle tree and its neighboring anchor merkle trees.
     /// @param nullifierHash: Nullifier hash.
     /// @param externalNullifier: External nullifier.
     /// @param proof: Zero-knowledge proof.
     /// @param verifier: Verifier address.
     function _verifyProof(
```
Comment (drewstone) [inline contracts/base/SemaphoreCore.sol:28]: I think we need to add some validation on the provided roots still. It doesn't look like we are checking that each root exists in the LinkableTreeData history. Wdyt @semaraugusto ?

**Review by drewstone — COMMENTED** [review]
Verifier names seem a bit odd too, presumably you're still updating all this @semaraugusto ?

File: contracts/Semaphore.sol (hunk 3 above)
Comment (semaraugusto) [reply contracts/Semaphore.sol:13]: no need for a double map bc the selector deals with the max-edge verifier selection

File: contracts/base/SemaphoreCore.sol (hunk 6 above)
Comment (semaraugusto) [reply contracts/base/SemaphoreCore.sol:28]: hm I had not thought about that, yeah, you're right. Current implementation allows claim of arbitrary merkle-roots not necessarily in the contract which is completely insecure

Comment (semaraugusto) [reply contracts/base/SemaphoreCore.sol:28]: tbh I don't even think the roots should be supplied to the verifyProof. We should just take it directly from the contract instead, no?

File: contracts/base/LinkableIncrementalBinaryTree.sol [hunk 7]
```diff
@@ -0,0 +1,399 @@
...
+
+        self.depth = depth;
+        self.roots[0] = zeros(depth-1);
+        self.maxEdges = maxEdges;
+    }
+
+    /// @dev Inserts a leaf in the tree.
+    /// @param self: Tree data.
+    /// @param leaf: Leaf to be inserted.
+    function insert(LinkableIncrementalTreeData storage self, uint256 leaf) public {
```
Comment (semaraugusto) [inline contracts/base/LinkableIncrementalBinaryTree.sol:78]: should be private/internal

File: contracts/base/LinkableIncrementalBinaryTree.sol [hunk 8]
```diff
@@ -0,0 +1,399 @@
...
+    /// @param self: Tree data.
+    /// @param leaf: Leaf to be removed.
+    /// @param proofSiblings: Array of the sibling nodes of the proof of membership.
+    /// @param proofPathIndices: Path of the proof of membership.
+    function remove(
+        LinkableIncrementalTreeData storage self,
+        uint256 leaf,
+        uint256[] calldata proofSiblings,
+        uint8[] calldata proofPathIndices
+    ) public {
```
Comment (semaraugusto) [inline contracts/base/LinkableIncrementalBinaryTree.sol:112]: also should be private/internal

**Review by semaraugusto — COMMENTED** [review]
adding the input from product meeting into semaphore

File: contracts/Semaphore.sol (hunk 3 above)
Comment (drewstone) [reply contracts/Semaphore.sol:13]: Oh right, so this verifier is a meta-verifier.

File: contracts/base/SemaphoreCore.sol (hunk 6 above)
Comment (drewstone) [reply contracts/base/SemaphoreCore.sol:28]: No, we shouldn't take directly from the contract as that introduces potential for race conditions (people making proofs while others are inserting).

File: contracts/base/LinkableIncrementalBinaryTree.sol (hunk 7 above)
Comment (drewstone) [reply contracts/base/LinkableIncrementalBinaryTree.sol:78]: Ok

File: contracts/base/LinkableIncrementalBinaryTree.sol (hunk 8 above)
Comment (drewstone) [reply contracts/base/LinkableIncrementalBinaryTree.sol:112]: Ok

**Pull request merged**